File: py/kintsugi/parsers.py
"""Turning raw tool output into typed findings.

A finding is only as good as its anchor: a file, a line, a machine code and
a message that the healer can act on. Output that cannot name a file inside
the source root is not a finding — it is noise, and a loop that heals noise
is how an auto-fixer starts breaking things.

This is a faithful port of the TypeScript engine's `strict` and `lines`
parsers — the two contracts the non-Node path (pytest, ruff, go test,
go vet, custom checks) speaks.
"""

import os
import re

from .fingerprint import fingerprint

# Tool noise that must never be treated as repo code: vendored deps, build
# output, VCS internals, virtualenvs and their package dirs, and the caches
# Python toolchains drop inside the repo.
SKIP = re.compile(
    r"(^|[\\/])(node_modules|dist|build|\.git|\.venv|venv|site-packages|"
    r"dist-packages|__pycache__|\.pytest_cache|\.mypy_cache|\.ruff_cache|"
    r"\.tox)([\\/]|$)"
)

_DRIVE = re.compile(r"^[A-Za-z]:[\\/]")


def is_abs_path(p: str) -> bool:
    # `C:/…` is absolute on Windows — and tool output can carry Windows
    # paths on any OS, so treat a drive-letter prefix as absolute everywhere.
    return os.path.isabs(p) or bool(_DRIVE.match(p))


def normalize_path(p: str, cwd: str) -> str | None:
    """Normalize a reported path against the source root; drop anything
    outside it. Returns forward-slash absolute paths so the rest of the
    engine never has to care which OS produced them."""
    cleaned = p.strip().replace("\\", "/")
    if not cleaned:
        return None

    # `file:///C:/repo/x.ts` → `C:/repo/x.ts`; `file:///repo/x.ts` →
    # `/repo/x.ts`. Stripping `file://` wholesale would drop the root slash
    # on POSIX and turn an absolute path into a relative one.
    if cleaned.startswith("file://"):
        rest = cleaned[len("file://"):]
        if re.match(r"^/[A-Za-z]:", rest):
            rest = rest[1:]
        abs_p = rest if is_abs_path(rest) else os.path.abspath(os.path.join(cwd, rest))
    else:
        abs_p = cleaned if is_abs_path(cleaned) else os.path.abspath(os.path.join(cwd, cleaned))
    abs_p = abs_p.replace("\\", "/")

    try:
        rel = os.path.relpath(abs_p, cwd).replace("\\", "/")
    except ValueError:
        # Different drive on Windows — the path is outside the root.
        return None
    if rel.startswith("..") or os.path.isabs(rel):
        return None
    if SKIP.search(rel):
        return None
    return abs_p


# ------------------------------------------------------------- strict

# Optional drive letter, then a path token, then `:line`, optional `:col`,
# then the message. The path must look like a file (an extension or a
# separator) and never contain whitespace, so `FAIL\texample.com/x\t0.02s`
# and `1 failed in 0.12s` cannot become phantom findings.
_STRICT_RE = re.compile(r"^\s*(?:(?:[A-Za-z]:)?([^:\s]+):(\d+)(?::(\d+))?:\s*(.*))$")
_FILEISH = re.compile(r"\.\w+$|[\\/]")
_FRAME = re.compile(r"^in [\w.]+$")
_WARNING = re.compile(r"^[A-Za-z][\w.]*Warning:\s")
_CODE = re.compile(r"^([A-Z]+\d+)\b")


def parse_strict(output: str, cwd: str, check: str):
    """The zero-config contract for non-TS toolchains — pytest, go test,
    go vet, ruff. Every finding must be anchored to a `file:line` inside the
    source root; unlike `lines`, a bare line is never a finding.

    Shapes read:

      pytest --tb=line   test_pricing.py:7: assert 8 == 10
      ruff               src/foo.py:12:5: F401 'os' imported but unused
      go test            foo_test.go:25: expected 10, got 5
      go vet             ./foo.go:12:2: fmt.Println is unused
    """
    findings = []
    for raw in output.split("\n"):
        # Trim CR/LF and padding: tool output on Windows is CRLF, and a
        # trailing `\r` defeats the `$` anchor.
        m = _STRICT_RE.match(raw.strip())
        if not m:
            continue
        head = m.group(1)
        if not _FILEISH.search(head):
            continue
        file = normalize_path(head, cwd)
        if not file:
            continue
        line = int(m.group(2))
        col = int(m.group(3)) if m.group(3) else None
        message = m.group(4).strip()
        # A traceback frame (`test_x.py:7: in test_tax_rate`) is scaffolding;
        # the error line that follows it carries the defect.
        if _FRAME.match(message):
            continue
        # pytest's `warnings summary` block lists `path:line:
        # CategoryWarning: message` lines in the same shape as findings — but
        # they are not the failure the check is reporting, so a noisy suite
        # must not surface them as phantom defects.
        if _WARNING.match(message):
            continue
        # Tool codes travel inside the message for some tools (ruff: `F401 [*]
        # 'os' imported but unused`) — lift the leading code token out so
        # rules can dispatch on it the way they do on TSxxxx.
        code_m = _CODE.match(message)
        code = code_m.group(1) if code_m else None
        evidence = {"message": message}
        if col is not None:
            evidence["col"] = col
        if code:
            evidence["code"] = code
        findings.append({
            "fingerprint": fingerprint(check, file, code or "", message),
            "check": check,
            "severity": "minor",
            "summary": message,
            "file": file,
            "line": line,
            "code": code,
            "evidence": evidence,
        })
    return findings


# ------------------------------------------------------------- lines

_LINES_RE = re.compile(r"^(?:([^:]+):(\d+:)?\s*)?(.*)$")


def parse_lines(output: str, cwd: str, check: str):
    """The plain-text contract for custom checks. Each non-empty line is
    either `path[:line]: message` or a bare message; a `path` must look like
    one (a dot-extension or a separator), so a message like `version-drift:
    …` is not mistaken for a file named `version-drift`."""
    findings = []
    for raw in output.split("\n"):
        line = raw.strip()
        if not line:
            continue
        m = _LINES_RE.match(line)
        head = m.group(1)
        looks_like_file = (
            bool(head)
            and " " not in head
            and (_FILEISH.search(head) is not None)
        )
        file = normalize_path(head, cwd) if looks_like_file else None
        line_no = int(m.group(2)[:-1]) if looks_like_file and m.group(2) else None
        message = (m.group(3) if looks_like_file else line).strip()
        if not message:
            continue
        findings.append({
            "fingerprint": fingerprint(check, file, "", message),
            "check": check,
            "severity": "minor",
            "summary": message,
            "file": file,
            "line": line_no,
            "evidence": {"message": message},
        })
    return findings

File: src/parsers.ts
import { resolve, isAbsolute, relative } from 'node:path';
import type { Finding } from './types.js';
import { fingerprint } from './fingerprint.js';

/**
 * Turning raw tool output into typed findings.
 *
 * A finding is only as good as its anchor: a file, a line, a machine code and
 * a message that the healer can act on. Output that cannot name a file inside
 * the source root is not a finding — it is noise, and a loop that heals noise
 * is how an auto-fixer starts breaking things.
 */

const SKIP = /(^|[\\/])(node_modules|dist|build|\.git)([\\/]|$)/;

/** Resolve a reported path against the source root; drop anything outside. */
function normalizePath(p: string, cwd: string): string | undefined {
  const cleaned = p.trim().replace(/^file:\/\//, '').replace(/\\/g, '/');
  if (!cleaned) return undefined;
  const abs = (isAbsolute(cleaned) ? cleaned : resolve(cwd, cleaned)).replace(/\\/g, '/');
  const rel = relative(cwd, abs).replace(/\\/g, '/');
  if (rel.startsWith('..') || SKIP.test(rel)) return undefined;
  return abs;
}

// ------------------------------------------------------------- tsc

/** `file(line,col): error TS2307: Cannot find module '...'` */
export function parseTsc(output: string, cwd: string, check: string): Finding[] {
  const findings: Finding[] = [];
  const re = /^(.+?)\((\d+),(\d+)\):\s*error\s+(TS\d+):\s*(.+)$/gm;
  for (const m of output.matchAll(re)) {
    const file = normalizePath(m[1], cwd);
    if (!file) continue;
    const code = m[4];
    const message = m[5].trim();
    findings.push({
      fingerprint: fingerprint(check, m[1], code, message),
      check,
      severity: 'major',
      summary: `${code}: ${message}`,
      file,
      line: Number(m[2]),
      code,
      evidence: { code, message, ...extractTscEvidence(code, message) },
    });
  }
  return findings;
}

function extractTscEvidence(code: string, message: string): Record<string, unknown> {
  if (code === 'TS6133') {
    const sym = message.match(/'([^']+)' is declared/);
    return sym ? { symbol: sym[1] } : {};
  }
  if (code === 'TS2307') {
    const mod = message.match(/Cannot find module '([^']+)'/);
    return mod ? { module: mod[1] } : {};
  }
  if (code === 'TS2305' || code === 'TS2459') {
    const member = code === 'TS2305'
      ? message.match(/has no exported member '([^']+)'/)?.[1]
      : message.match(/declares '([^']+)' locally/)?.[1];
    // tsc wraps the module path in *additional* quotes inside the message:
    // Module '"/abs/pricing.ts"' has no exported member 'lineTotal'.
    // TS2459 instead reports the specifier as written ('./pricing.js').
    const mod = message.match(/Module '(\"?)([^'\"]+)\1'/);
    return {
      ...(member ? { member } : {}),
      ...(mod ? { module: mod[2] } : {}),
    };
  }
  return {};
}

// ------------------------------------------------------------- tap

/**
 * Node's built-in test runner (TAP output), as produced by `node --test` /
 * `tsx --test`. A failed test is a block:
 *
 *   not ok 1 - applyTax applies the 10% tax rate
 *     ---
 *     AssertionError: expected 8 to equal 10
 *         at file:///…/test/pricing.test.ts:10:23
 */
export function parseTap(output: string, cwd: string, check: string): Finding[] {
  const findings: Finding[] = [];
  const lines = output.split('\n');
  // Node 22 indents the per-file TAP block, so the anchors must tolerate
  // leading whitespace.
  const notOk = /^\s*not ok\s+\d+\s+-\s+(.+)$/;
  // A failed test block names its location either as a stack line or as the
  // TAP `location:` field.
  const atRe = /^\s*at\s+(?:file:\/\/\/?)?(.+?):(\d+):\d+\)?/;
  const locationRe = /^\s*location:\s*'(.+?):(\d+):(\d+)'/;

  for (let i = 0; i < lines.length; i++) {
    const m = lines[i].match(notOk);
    if (!m) continue;
    const title = m[1].trim();
    // The TAP reporter wraps each test file in its own block; a file whose
    // tests fail also gets a file-level "not ok". That is a harness line,
    // not a defect — skip it, or every broken test reports twice.
    if (/\.(test|spec)\.[cm]?[jt]sx?$/.test(title)) continue;

    let file: string | undefined;
    let line: number | undefined;
    for (let j = i + 1; j < Math.min(i + 25, lines.length); j++) {
      const at = lines[j].match(atRe);
      if (at) {
        file = normalizePath(at[1], cwd);
        line = Number(at[2]);
        break;
      }
      const loc = lines[j].match(locationRe);
      if (loc) {
        file = normalizePath(loc[1], cwd);
        line = Number(loc[2]);
        break;
      }
      // Next test block without a location — stop hunting.
      if (j > i + 1 && /^\s*(ok|not ok)\b/.test(lines[j])) break;
    }

    findings.push({
      fingerprint: fingerprint(check, file, '', title),
      check,
      severity: 'blocker',
      summary: title,
      file,
      line,
      evidence: { title },
    });
  }
  return findings;
}

// ------------------------------------------------------------- lines

/**
 * The plain-text contract for custom checks. Each non-empty line is either
 * `path:line: message` or a bare message; a `path` must look like one (a
 * dot-extension or a separator), so a message like `version-drift: …` is not
 * mistaken for a file named `version-drift`.
 */
export function parseLines(output: string, cwd: string, check: string): Finding[] {
  const findings: Finding[] = [];
  for (const raw of output.split('\n')) {
    const line = raw.trim();
    if (!line) continue;
    const m = line.match(/^(?:([^:]+):(\d+):\s*)?(.*)$/);
    const head = m?.[1];
    const looksLikeFile =
      !!head && !head.includes(' ') && (/\.\w+$/.test(head) || /[\\/]/.test(head));
    const file = looksLikeFile ? normalizePath(head!, cwd) : undefined;
    const lineNo = looksLikeFile && m?.[2] ? Number(m[2]) : undefined;
    const message = (looksLikeFile ? m?.[3] : line).trim();
    if (!message) continue;

    findings.push({
      fingerprint: fingerprint(check, file, '', message),
      check,
      severity: 'minor',
      summary: message,
      file,
      line: lineNo,
      evidence: { message },
    });
  }
  return findings;
}
